### PR TITLE
Includes basic np dlpack function that is tested to work across jax a…

### LINKED ISSUE #1063: use torch dlpack instead of cupy to convert cuda array to numpy

### PR DESCRIPTION
**Pull request by apasarkar**
Fixes #1063

The numpy dlpack fully sidesteps the need to rely on torch, but it requires numpy > 2.1. I think this is ok, since numpy is now at 2.5, and only users who care about high-performance (i.e. computing on GPUs) will use this execution path anyways.

### CODE REVIEW

**Comment by apasarkar** [comment]
@kushalkolar so one issue I've been seeing is that the cuda array interface execution path seems to produce spatially transposed data. See below for a reproducible example.

```

## Make a reproducible example:
data = np.random.rand(100, 50, 200)
data_torch = torch.as_tensor(data)

ref_ranges = {'time': (0, data.shape[0], 1)}
subplot_names = ['data_numpy', 'data_torch']
curr_ndw = fpl.NDWidget(ref_ranges,
                        names=subplot_names,
                        shape=(1, 2),
                        size=(800, 800))

dims = ('time', 'height', 'width')
spatial_dims = ('height', 'width')

ndw_vid_numpy = curr_ndw['data_numpy'].add_nd_image(data,
                                       dims,
                                       spatial_dims,
                                       name = 'data_numpy')

ndw_vid_torch = curr_ndw['data_torch'].add_nd_image(data_torch,
                                       dims,
                                       spatial_dims,
                                       name = 'data_torch')



curr_ndw.show()
```

The reason for this is because at the following line, the transpose does different things if the input is a tensor vs. if it is a numpy array. https://github.com/fastplotlib/fastplotlib/blob/91faf5b16244c7e44bd1d07828744a71716d3458/fastplotlib/widgets/nd_widget/_base.py#L537

For a tensor, my_tensor.transpose(0, 1) will swap the axes 0 and 1. Otoh, my_numpy_array.transpose(0, 1) will keep the desired shape. 

Is there any reason we are doing this final transpose step? For an individual ndprocessor, it does not seem necessary. I think there are two options: 
(1) If there's no reason, drop it. 
(2) If we need to do it, defer this to after we convert the dlpack-compatible array to numpy. This way the function call will be consistent.

File: fastplotlib/utils/functions.py [hunk 1]
```diff
@@ -408,14 +408,10 @@ def parse_cmap_values(
...
-    except ImportError:
-        raise ImportError(
-            "`cupy` is required to work with GPU arrays\npip install cupy"
-        )
-
-    return cupy.asnumpy(arr)
+
+    data = np.from_dlpack(arr, device='cpu') #This requires numpy >= 2.1
+    return data
+
```
Comment (kushalkolar) [inline fastplotlib/utils/functions.py:414]: I just tried this and it works, nice!

File: fastplotlib/utils/functions.py [hunk 2]
```diff
@@ -408,14 +408,10 @@ def parse_cmap_values(
...
-    try:
-        import cupy
-    except ImportError:
-        raise ImportError(
-            "`cupy` is required to work with GPU arrays\npip install cupy"
-        )
-
-    return cupy.asnumpy(arr)
+
+    data = np.from_dlpack(arr, device='cpu') #This requires numpy >= 2.1
```
Comment (kushalkolar) [inline fastplotlib/utils/functions.py]:
```suggestion
    data = np.from_dlpack(arr, device='cpu')
```

**Review by kushalkolar — COMMENTED** [review]
Where does it say that the `from_dlpack` requires numpy >= 2.1? Anyways that's fine, v2.1 is 2 years old at this point. pygfx requires min 2.1 anyways too https://github.com/pygfx/pygfx/blob/main/pyproject.toml#L28C1-L28C55

**Comment by kushalkolar** [comment]
> For a tensor, my_tensor.transpose(0, 1) will swap the axes 0 and 1. Otoh, my_numpy_array.transpose(0, 1) will keep the desired shape.
> 
> Is there any reason we are doing this final transpose step? For an individual ndprocessor, it does not seem necessary. I think there are two options: (1) If there's no reason, drop it. (2) If we need to do it, defer this to after we convert the dlpack-compatible array to numpy. This way the function call will be consistent.

Annoying that torch behavior is different from numpy. 

The transpose is necessary by definition, that's how the NDWidget spec is. The specified spatial dims order defines the dim order for the graphic data.

Doing the transpose after it's converted to numpy sounds like the way to go.

**Comment by apasarkar** [comment]
@kushalkolar Updated this -- so I've delayed the permutation operation until the sliced data is guaranteed to be a numpy array. nd vectors processor, nd image processor, and nd positions processor all subclass ndprocessor, so those are the files where i made this change. 

One thing to note: with this setup, the spatial functions must be written so they are applied data that has not yet been permuted. for example, if i specify data like this: 

dims = ['time', 'a', 'b']
spatial_dims = ['b', 'a'] # This means we end up transposing the sliced data

The spatial window function received the spatial dimensions in order (a, b) with these changes. 

I actually think this is ok -- this forces the user to write all their window/spatial function logic using the dimension order specified by dims, but the final order is transposed based on what's in spatial_dims. So we would ofc need to update the specs. I think so long as the user can write their whole pipeline (raw data --> displayed graphic) with a fixed dimension ordering it will be clear. 

Let me know what you think!

**Comment by kushalkolar** [comment]
> One thing to note: with this setup, the spatial functions must be written so they are applied data that has not yet been permuted. for example, if i specify data like this:
> 
> dims = ['time', 'a', 'b'] spatial_dims = ['b', 'a'] # This means we end up transposing the sliced data
> 
> The spatial window function received the spatial dimensions in order (a, b) with these changes.
> 
> I actually think this is ok -- this forces the user to write all their window/spatial function logic using the dimension order specified by dims, but the final order is transposed based on what's in spatial_dims. So we would ofc need to update the specs. I think so long as the user can write their whole pipeline (raw data --> displayed graphic) with a fixed dimension ordering it will be clear.
> 
> Let me know what you think!

I think that makes sense. The whole point is that window_funcs and spatial_funcs should be fast and async, and things are naturally async when they remain in torch as long as possible. The `spatial_dims` arg for an `NDGraphic` is really just supposed to be the _display order_ of the dimensions for rendering the graphic, the `[x, y, z]` dims for positional data, `[row, col]` for image data nad `[depth, row, col]` for volumetric image data. The user needs to know what their dims are to create their window_funcs -> spatial_funcs pipeline.

Maybe we should rename the arg to `display_order` or `display_dims` or `render_dims` or something?

**Comment by apasarkar** [comment]
ordered_spatial_dims might be a good one, specifying the two roles of this parameter: (1) it specifies what the spatial dims are and (2) it gives an order to output the spatial dims. thoughts?

**Comment by apasarkar** [comment]
@kushalkolar I'll also flag the following potential issue: 

https://github.com/fastplotlib/fastplotlib/blob/91faf5b16244c7e44bd1d07828744a71716d3458/fastplotlib/widgets/nd_widget/_base.py#L525

At this line, a function "squeeze" is used. On a first read, I expected it to fail with torch tensors, since the documented parameters are tensor.squeeze(dim = ...), whereas the code is doing .squeeze(axis =...). Turns out at some point in pytorch the developers allowed axis to make it interoperable with numpy. 

Anyways it's not clear with other libraries that are compatible with dlpack whether this functionality will break.

**Comment by kushalkolar** [comment]
`cupy` is supposed to be just like numpy and accepts `axis`: https://docs.cupy.dev/en/stable/reference/generated/cupy.squeeze.html
`jax` likewise: https://docs.jax.dev/en/latest/_autosummary/jax.numpy.squeeze.html

where is it documented that torch also accepts `axis`? I don't see it here: https://docs.pytorch.org/docs/2.13/generated/torch.squeeze.html#torch.squeeze

**Comment by apasarkar** [comment]
@kushalkolar  Example:

https://discuss.pytorch.org/t/confuse-about-how-torch-argmax-supporting-axis-parameter-inplace-of-dim/182592

Looks like under the hood there is some code that parses keywords.

**Comment by kushalkolar** [comment]
> @kushalkolar Example:
> 
> https://discuss.pytorch.org/t/confuse-about-how-torch-argmax-supporting-axis-parameter-inplace-of-dim/182592
> 
> Looks like under the hood there is some code that parses keywords.

wow that's hidden deep

**Comment by apasarkar** [comment]
@kushalkolar I'm thinking from the standpoint of documentation. We ideally want to be able to tell someone "your dlpack compliant array can work here". Maybe there is not a workaround with the axis squeeze thing though

**Comment by kushalkolar** [comment]
If it works with torch, cupy and Jax that covers most use cases I'd think?

**Comment by apasarkar** [comment]
@kushalkolar  Fair enough, we can just make sure to note this in the documentation. 

Ok so what's left here - just the rename of the spatial_dims param to something else? If you're ok with renaming is ordered_spatial_dims, we'll have to update many of the function signatures in ndwidget, so let's make sure we're both happy with the naming here and I can go ahead and do it.

**Comment by kushalkolar** [comment]
I added it to the list of renames to do later in bulk on #971

**Comment by kushalkolar** [comment]
what's left is the suggestion and updating pyproject.toml and merge latest `ndwidget` into this

File: fastplotlib/utils/functions.py (hunk 2 above)
Comment (apasarkar) [reply fastplotlib/utils/functions.py]: Done!

**Comment by apasarkar** [comment]
@kushalkolar Added the suggested change. 

Re: pyproject.toml, we are now using np.from_dlpack exclusively to bring gpu arrays to cpu and from there do zero-cost conversion to numpy. So it looks like we don't need to change the pyproject.toml at all.

**Comment by kushalkolar** [comment]
> @kushalkolar Added the suggested change. 
> 
> Re: pyproject.toml, we are now using np.from_dlpack exclusively to bring gpu arrays to cpu and from there do zero-cost conversion to numpy. So it looks like we don't need to change the pyproject.toml at all. 

For the min version pin?

**Comment by apasarkar** [comment]
@kushalkolar sorry I thought I did that earlier haha. 
I think g2g in latest commit.

File: fastplotlib/widgets/nd_widget/_nd_image.py [hunk 3]
```diff
@@ -137,6 +137,11 @@ def __init__(
         self._compute_histogram = compute_histogram
         self._recompute_histogram()
 
+        # Axis order of the spatial dimensions to display
+        self._spatial_dims_int = tuple(
+            self.spatial_dims.index(d) for d in self.dims if d in self.spatial_dims
+        )
+
```
Comment (kushalkolar) [inline fastplotlib/widgets/nd_widget/_nd_image.py]: spatial_dims is a mutable property, so this should be a read only property or set in the spatial_dims setter as a private attribute (probably preferable, it doesn't have to be public) and not created once in the constructor

**Review by kushalkolar — CHANGES_REQUESTED** [review]

File: fastplotlib/widgets/nd_widget/_nd_image.py (hunk 3 above)
Comment (apasarkar) [reply fastplotlib/widgets/nd_widget/_nd_image.py]: @kushalkolar good point, updated in latest commit. I also renamed spatial_dims_int to spatial_dims_indices for clarity

File: fastplotlib/widgets/nd_widget/_base.py [hunk 4]
```diff
@@ -198,6 +198,19 @@ def spatial_dims(self, sdims: Sequence[str]):
...
+        self._spatial_dims_indices = tuple(
+            self.spatial_dims.index(d) for d in self.dims if d in self.spatial_dims
+        )
+
+    @property
+    def spatial_dims_indices(self) -> tuple[int, ...]:
+        """
+        The ordered sequence of data indices that will be displayed
+        """
+        return self._spatial_dims_indices
```
Comment (kushalkolar) [inline fastplotlib/widgets/nd_widget/_base.py]:
```suggestion
        return tuple(
            self.spatial_dims.index(d) for d in self.dims if d in self.spatial_dims
        )
```

File: fastplotlib/widgets/nd_widget/_base.py [hunk 5]
```diff
@@ -198,6 +198,19 @@ def spatial_dims(self, sdims: Sequence[str]):
...
 
+        ## This is the ordered sequence of data indices that will be displayed
+        self._spatial_dims_indices = tuple(
+            self.spatial_dims.index(d) for d in self.dims if d in self.spatial_dims
+        )
+
+    @property
+    def spatial_dims_indices(self) -> tuple[int, ...]:
+        """
+        The ordered sequence of data indices that will be displayed
```
Comment (kushalkolar) [inline fastplotlib/widgets/nd_widget/_base.py]:
```suggestion
        ordered spatial dim indices that correspond to the named spatial dims
```

File: fastplotlib/widgets/nd_widget/_base.py [hunk 6]
```diff
@@ -198,6 +198,19 @@ def spatial_dims(self, sdims: Sequence[str]):
 
         self._spatial_dims = tuple(sdims)
 
+        ## This is the ordered sequence of data indices that will be displayed
+        self._spatial_dims_indices = tuple(
+            self.spatial_dims.index(d) for d in self.dims if d in self.spatial_dims
+        )
+
```
Comment (kushalkolar) [inline fastplotlib/widgets/nd_widget/_base.py]:
```suggestion
```

File: fastplotlib/widgets/nd_widget/_nd_positions/_nd_positions.py [hunk 7]
```diff
@@ -112,6 +112,7 @@ def __init__(
         self.cmap_transform_each = cmap_transform_each
         self.sizes = sizes
 
+
```
Comment (kushalkolar) [inline fastplotlib/widgets/nd_widget/_nd_positions/_nd_positions.py:115]:
```suggestion
```

File: fastplotlib/utils/functions.py [hunk 8]
```diff
@@ -408,14 +408,10 @@ def parse_cmap_values(
...
 def cuda_to_numpy(arr: CudaArrayProtocol) -> np.ndarray:
-    try:
-        import cupy
-    except ImportError:
-        raise ImportError(
-            "`cupy` is required to work with GPU arrays\npip install cupy"
-        )
-
-    return cupy.asnumpy(arr)
+
```
Comment (kushalkolar) [inline fastplotlib/utils/functions.py:411]:
```suggestion
```

File: fastplotlib/widgets/nd_widget/_nd_positions/_nd_positions.py [hunk 9]
```diff
@@ -298,6 +299,18 @@ def spatial_dims(self, sdims: tuple[str, str, str]):
...
+            self.spatial_dims.index(d) for d in self.dims if d in self.spatial_dims
+        )
+
+    @property
+    def spatial_dims_indices(self) -> tuple[int, ...]:
+        """
+        The ordered sequence of data indices that will be displayed
+        """
+        return self._spatial_dims_indices
+
```
Comment (kushalkolar) [inline fastplotlib/widgets/nd_widget/_nd_positions/_nd_positions.py]:
```suggestion
```

File: fastplotlib/widgets/nd_widget/_nd_positions/_nd_positions.py [hunk 10]
```diff
@@ -555,6 +568,8 @@ async def get(self, indices: dict[str, Any]) -> dict[str, ArrayProtocol]:
         if isinstance(data, CudaArrayProtocol):
             data = await run_in_thread_pool(self._executor, cuda_to_numpy, data)
 
+        data = data.transpose(*self._spatial_dims_int)
```
Comment (kushalkolar) [inline fastplotlib/widgets/nd_widget/_nd_positions/_nd_positions.py]:
```suggestion
        data = data.transpose(*self.spatial_dims_indices)
```

you made it a public property but didn't use it?

File: fastplotlib/widgets/nd_widget/_base.py [hunk 11]
```diff
@@ -198,6 +198,19 @@ def spatial_dims(self, sdims: Sequence[str]):
...
+    @property
+    def spatial_dims_indices(self) -> tuple[int, ...]:
+        """
+        The ordered sequence of data indices that will be displayed
+        """
+        return self._spatial_dims_indices
+
+
```
Comment (kushalkolar) [inline fastplotlib/widgets/nd_widget/_base.py:207]:
```suggestion
```

File: fastplotlib/widgets/nd_widget/_base.py
Comment (kushalkolar) [inline fastplotlib/widgets/nd_widget/_base.py:531]: I think you can use the public property and make sure the rest of these code blocks are valid now?

File: fastplotlib/widgets/nd_widget/_nd_image.py [hunk 12]
```diff
@@ -186,6 +186,18 @@ def spatial_dims(self, sdims: tuple[str, str] | tuple[str, str, str]):
...
+    @property
+    def spatial_dims_indices(self) -> tuple[int, ...]:
+        """
+        The ordered sequence of data indices that will be displayed
+        """
+        return self._spatial_dims_indices
+
```
Comment (kushalkolar) [inline fastplotlib/widgets/nd_widget/_nd_image.py]:
```suggestion
```

File: fastplotlib/widgets/nd_widget/_nd_image.py [hunk 13]
```diff
@@ -258,7 +270,8 @@ async def get(self, indices: dict[str, Any]) -> ArrayProtocol:
         if isinstance(window_output, CudaArrayProtocol):
             window_output = await run_in_thread_pool(self._executor, cuda_to_numpy, window_output)
 
-        return window_output
+
```
Comment (kushalkolar) [inline fastplotlib/widgets/nd_widget/_nd_image.py]:
```suggestion
```

File: fastplotlib/widgets/nd_widget/_nd_vectors.py [hunk 14]
```diff
@@ -143,6 +143,18 @@ def spatial_dims(self, sdims: tuple[str, str, str]):
...
+            self.spatial_dims.index(d) for d in self.dims if d in self.spatial_dims
+        )
+
+    @property
+    def spatial_dims_indices(self) -> tuple[int, ...]:
+        """
+        The ordered sequence of data indices that will be displayed
+        """
+        return self._spatial_dims_indices
+
```
Comment (kushalkolar) [inline fastplotlib/widgets/nd_widget/_nd_vectors.py]:
```suggestion
```

File: fastplotlib/widgets/nd_widget/_nd_vectors.py [hunk 15]
```diff
@@ -175,7 +187,8 @@ async def get(self, indices: dict[str, Any]) -> ArrayProtocol:
         if isinstance(window_output, CudaArrayProtocol):
             window_output = await run_in_thread_pool(self._executor, cuda_to_numpy, window_output)
 
-        return window_output
+
```
Comment (kushalkolar) [inline fastplotlib/widgets/nd_widget/_nd_vectors.py]:
```suggestion
```

File: fastplotlib/widgets/nd_widget/_nd_vectors.py [hunk 16]
```diff
@@ -175,7 +187,8 @@ async def get(self, indices: dict[str, Any]) -> ArrayProtocol:
...
+        return window_output.transpose(*self._spatial_dims_indices)
```
Comment (kushalkolar) [inline fastplotlib/widgets/nd_widget/_nd_vectors.py]:
```suggestion
        return window_output.transpose(*self.spatial_dims_indices)
```

it's a public property use it

**Review by kushalkolar — CHANGES_REQUESTED** [review]

**Comment by apasarkar** [comment]
@kushalkolar thanks -- all makes sense. made the changes, spatial_dims_indices is computed and set in the public property of the base class now. i was hesitant to do it this way at the start because the computation to define spatial dims indices will now be run at every single getitem. but if the number of dimensions is small, the overhead should never be a problem.

**Comment by kushalkolar** [comment]
It's probably a nanosecond scale operation.

File: fastplotlib/widgets/nd_widget/_base.py
Comment (kushalkolar) [reply fastplotlib/widgets/nd_widget/_base.py:531]: Oh wait this is different nevermind (but similar name lol), this is to remove all non-spatial dims from the window slice.

Comment (apasarkar) [reply fastplotlib/widgets/nd_widget/_base.py:531]: yep!
